ext/DatadogArena/arena.c

```diff
@@ -0,0 +1,59 @@
+#include "datadog/arena.h"
+
+#include <stdlib.h>
+
+_Static_assert(sizeof(size_t) == 8, "Unexpected size_t size");
+
+extern inline char *datadog_arena_alloc(datadog_arena **arena_ptr, size_t size);
+extern inline bool datadog_arena_try_alloc(datadog_arena *arena, size_t size, char **result);
+
+/* prefer powers of 2 */
+datadog_arena *datadog_arena_create(size_t size) {
+    datadog_arena *arena = (datadog_arena *)calloc(1, size);
+
+    size_t aligned = DATADOG_ARENA_ALIGNED_SIZE(sizeof(datadog_arena));
+    arena->ptr = (char *)arena + aligned;
+    arena->end = (char *)arena + size;
+    arena->prev = NULL;
+
+    return arena;
+}
+
+void datadog_arena_grow(datadog_arena **arena_ptr, size_t min_size) {
+    datadog_arena *arena = *arena_ptr;
+    size_t prev_size = arena->end - (char *)arena;
+    min_size += DATADOG_ARENA_ALIGNED_SIZE(sizeof(datadog_arena));
+    size_t size = prev_size > min_size ? prev_size : min_size;
+    // todo: align min_size to a power of 2 if it doesn't fit?
+
+    datadog_arena *new_arena = datadog_arena_create(size);
+    new_arena->prev = *arena_ptr;
+    *arena_ptr = new_arena;
+}
+
+void datadog_arena_destroy(datadog_arena *arena) {
+    do {
+        datadog_arena *prev = arena->prev;
+        free(arena);
+        arena = prev;
+    } while (arena);
+}
+
+char *datadog_arena_checkpoint(datadog_arena *arena) { return arena->ptr; }
+
+// todo: what to do when checkpoint isn't found in arena?
```

Comment:
Maybe in a follow up PR we can figure out syntax to designate "debug mode" across the shared/vendored components and then do an assertion for the checkpoint in debug mode. WDYT?

Reply:
On one hand having a debug build that switches behavior at compile-time is pretty nice as it removes checks from runtime, but it does mean you can't enable/disable it easily. I'm not sure.

Reply:
Reading the code again, it looks like a checkpoint that doesn't exist in the arena will invalidate the arena pointer. I think ideally the pointer would not change if the checkpoint did not exist and the caller can check the return value for an error state.

Reply:
Walking the entire arena to find it does make it more expensive. It's probably not a common operation so that may be okay.